### ISSUE TITLE
Pod scheduling conundrum

### ISSUE THREAD

**Issue by sheevy** [kubeflow/mpi-operator#553]
I have an MPIJob, with 48 worker replicas, each getting a single MPI rank. I have 3 nodes with 16 cores each. Is there a way to deploy the MPIJob such that pods with ranks 0-15 go onto the first machine, pods 16-31 to the second and pods 32-47 to thte third?

Thanks

**Comment by tenzen-y** [comment]
Hi, @sheevy. Thank you for creating this issue!

I think the kube-scheduler should be responsible for which node the Pods are scheduled, not the mpi-operator.

So you can set an appropriate [NodeAffinities](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity) to the `.spec.mpiReplicaSpecs.Launcher.template.spec.affinity` and the `.spec.mpiReplicaSpecs.Worker.template.spec.affinity` in the MPIJob.

**Comment by sheevy** [comment]
Thanks @tenzen-y I totally agree it's not mpi-operator responsibility. I'm just looking for help here as I'm k8s noob, and my problem arose when using mpi-operator.
Do you happen to have a MWE to demonstrate this, I've read the docs and it's still not clear how different workers from MPIJob can have diffferent affinities. Won't they have the same affinity?

**Comment by tenzen-y** [comment]
> Won't they have the same affinity?

Yes, you're understanding is right. All workers have the same affinity.

Do you want to restrict the number of pods for workers, which is scheduled per node, right? (in this case, 16?) 

https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/

If so, the PodTopologySpreadConstraints may be able to fulfill your requirements.

**Comment by sheevy** [comment]
Ok, I'll try setting skew=1 in hope they are evenly distributed.

In practise, I have more strict requirements, each group of 16 workers works on a single simulation so they communicate a lot and benefit from being collocated on a single machine. While between groups, they are just 3 different simulations, which just need to combine results after the sims are done. So ideally, I'd like to have control which 16 pods get scheduled onto first node (that would be pods working on the same simulation), next 16 pods working on second simulation would go onto the second node, an so forth.

**Comment by sheevy** [comment]
That didn't work, I'm closing the thread as it's not most approperiate forum to be discussing this.
Thanks for all the help!

**Comment by alculquicondor** [comment]
It would be very hard to provide this semantics to kube-scheduler.
Instead, I suggest you create bigger pods. Each pod can require 16 cores, and pair it with`slotsPerWorker: 16`.

**Comment by sheevy** [comment]
Thanks. That's actually what I was running at first, but for unexplained reason it failed for some of my jobs, so I tried to work around it. I will revisit that, and try to pull some relevant logs/info in hope t dig deeper why it didn't work.
